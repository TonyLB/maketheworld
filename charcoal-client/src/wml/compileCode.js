const compileCode = (src) => {
    src = 'with (sandbox) {' + src + '}'
    // eslint-disable-next-line no-new-func
    const code = new Function('sandbox', src)
  
    return function (sandbox) {
        const sandboxProxy = new Proxy(sandbox, {
            has: () => true,
            get: (target, key) => (key === Symbol.unscopables ? undefined: target[key])
        })
        return code(sandboxProxy)
    }
}

exports.compileCode = compileCode
