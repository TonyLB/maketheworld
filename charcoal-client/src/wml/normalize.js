import { produce } from 'immer'
import { v4 as uuidv4 } from 'uuid'

export class WMLNormalizeError extends Error {
    constructor(message) {
        super(`WMLNormalize: ${message}`)
        this.name = 'WMLNormalizeError'
    }
}

export class NormalizeTagMismatchError extends WMLNormalizeError {
    constructor(message) {
        super(message)
        this.name = 'TagMismatchError'
    }
}

export class NormalizeKeyMismatchError extends WMLNormalizeError {
    constructor(message) {
        super(message)
        this.name = 'KeyTagMismatchError'
    }
}

let generatedKeys = {}

export const clearGeneratedKeys = () => {
    generatedKeys = {}
}

const nextGeneratedKey = (tag) => {
    const currentValues = Object.keys(generatedKeys)
        .filter((key) => (key.startsWith(`${tag}-`)))
        .map((key) => (key.slice(`${tag}-`.length)))
    const maxValue = currentValues.reduce((previous, value) => {
        const numericValue = parseInt(value) || 0
        return (numericValue > previous) ? numericValue : previous
    }, 0)
    return `${tag}-${maxValue}`
}

const keyForValue = (tag, value) => {
    const [foundKey] = Object.entries(generatedKeys)
        .find(([key, checkValue]) => (
            key.startsWith(`${tag}-`) &&
            (checkValue === value)
        )) || [null]
    if (foundKey) {
        return foundKey
    }
    const syntheticKey = nextGeneratedKey(tag)
    generatedKeys[syntheticKey] = value
    return syntheticKey
}

const getCurrentAppearance = (existingMap = {}, key) => {
    if ((existingMap[key]?.appearances || []).length > 0) {
        return {
            appearance: existingMap[key].appearances[-1],
            index: existingMap[key].appearances.length - 1,
            valid: true
        }
    }
    return { valid: false }
}

const pullProperties = (node) => {
    return Object.entries(node).reduce((previous, [key, value]) => {
        let pullTags = [
            'tag',
            'key',
        ]
        switch(node.tag) {
            case 'Exit':
                pullTags = [...pullTags, 'to', 'from']
                break
            case 'Variable':
                pullTags.push('default')
                break
            case 'Action':
                pullTags.push('src')
                break
            case 'Asset':
                pullTags = [...pullTags, 'name', 'fileName', 'player', 'importMap', 'zone']
                break
        }
        if (pullTags.includes(key)) {
            return {
                ...previous,
                topLevel: {
                    ...previous.topLevel,
                    [key]: value
                }
            }
        }
        else {
            return {
                ...previous,
                appearance: {
                    ...previous.appearance,
                    [key]: value
                }
            }
        }
    }, { topLevel: {}, appearance: {} })
}

export const transformNode = (contextStack, node) => {
    if (node.tag === 'Exit') {
        const roomIndex = contextStack.reduceRight((previous, { tag }, index) => (((tag === 'Room') && (previous === -1)) ? index : previous), -1)
        if (roomIndex === -1) {
            return {
                contextStack,
                node: {
                    key: `${node.from}#${node.to}`,
                    ...node
                }
            }
        }
        const roomKey = contextStack[roomIndex].key
        const { to, from, ...rest } = node
        const returnValue = {
            key: `${from || roomKey}#${to || roomKey}`,
            to: to || roomKey,
            from: from || roomKey,
            ...rest
        }
        if (from && from !== roomKey) {
            const contextStackBeforeRoom = contextStack.slice(0, roomIndex)
            const contextStackAfterRoom = contextStack.slice(roomIndex + 1)
            //
            // NOTE: Context stacks within the sibling room element created do not have
            // an index item.  addElement will need to infer the correct index (either
            // the current for a matching element, or a created next wrapper)
            //
            return {
                contextStack: [
                    ...contextStackBeforeRoom,
                    {
                        key: from,
                        tag: 'Room'
                    },
                    ...(contextStackAfterRoom.map(({ key, tag }) => ({ key, tag })))
                ],
                node: returnValue
            }
        }
        else {   
            return {
                contextStack,
                node: returnValue
            }
        }
    }
    if (node.tag === 'Condition') {
        const key = keyForValue('Condition', node.src)
        return {
            node: {
                key,
                ...node    
            },
            contextStack
        }
    }
    return { node, contextStack }
}

//
// mergeElements tracks how to add an element into the normalized structure, given the contextStack in
// which it is encountered.
//
const mergeElements = (previous, contextStack, node) => {
    const deepEqual = (listA, listB) => {
        return JSON.stringify(listA) === JSON.stringify(listB)
    }

    const [currentAppearance] = previous.appearances.slice(-1)
    const { topLevel, appearance: incomingAppearance } = pullProperties(node)
    //
    // Integrate the topLevel data, making sure there are no conflicts
    //
    if (topLevel.key !== previous.key) {
        throw new NormalizeKeyMismatchError('Keys somehow mismatched in normalize')
    }
    if (topLevel.tag !== previous.tag) {
        throw new NormalizeTagMismatchError(`Key '${topLevel.key}' is used to define elements of different tags ('${previous.tag}' and '${topLevel.tag}')`)
    }

    //
    // If it is a sibling to a similarly keyed item, in the same context,
    // with no similarly-keyed items in other contexts (e.g. a conditional
    // redefining some properties after the first sibling) then its data
    // can be folded in ... sequential siblings get merged.
    //
    if (deepEqual(currentAppearance.contextStack, contextStack)) {
        const priorAppearances = previous.appearances.slice(0, -1)
        return {
            ...previous,
            ...topLevel,
            appearances: [
                ...priorAppearances,
                {
                    ...currentAppearance,
                    ...incomingAppearance,
                    contents: [
                        ...(currentAppearance.contents || []),
                        ...(incomingAppearance.contents || [])
                    ]
                }
            ]
        }
    }
    //
    // Otherwise, create a new appearance for the item
    //
    else {
        return {
            ...previous,
            ...topLevel,
            appearances: [
                ...previous.appearances,
                {
                    contextStack,
                    ...incomingAppearance
                }
            ]
        }
    }
}

export const addElement = (existingMap = {}, { contextStack = [], node }) => {
    const { key } = node
    const { contextFilledMap, filledContext } = contextStack.reduce((previous, { key, tag, index }) => {
        if (index !== undefined) {
            return {
                ...previous,
                filledContext: [...previous.filledContext, { key, tag, index }]
            }
        }
        else {
            const newMap = addElement(
                previous.contextFilledMap,
                {
                    contextStack: previous.filledContext,
                    node: {
                        key,
                        tag
                    }
                }
            )
            return {
                contextFilledMap: newMap,
                filledContext: [
                    ...previous.filledContext,
                    {
                        key,
                        tag,
                        index: newMap[key].appearances.length - 1
                    }
                ]
            }
        }
    }, { contextFilledMap: existingMap, filledContext: [] })
    return produce(contextFilledMap, (draftMap) => {
        let merged = false
        if (draftMap[key]) {
            //
            // If the item has already appeared, check whether to merge
            // with a same-level sibling, or add a new appearance
            //
            const priorAppearances = draftMap[key].appearances.length
            draftMap[key] = mergeElements(draftMap[key], filledContext, node)
            if (draftMap[key].appearances.length === priorAppearances) {
                merged = true
            }
        }
        else {
            //
            // If this is the first appearance of the item, add it to the
            // map fresh.
            //
            const { topLevel, appearance } = pullProperties(node)
            draftMap[key] = {
                ...topLevel,
                appearances: [{
                    contextStack: filledContext,
                    ...appearance
                }]
            }
        }
        //
        // If a new element is added in a context, the parent item needs to have its
        // contents updated to reflect that.
        //
        if (filledContext.length > 0 && !merged) {
            const { key, index } = filledContext.slice(-1)[0]
            const { index: nodeIndex } = getCurrentAppearance(draftMap, node.key)
            draftMap[key].appearances[index].contents = [
                ...(draftMap[key].appearances[index].contents || []),
                {
                    key: node.key,
                    tag: node.tag,
                    index: nodeIndex
                }
            ]
        }
    })
}

export const normalize = (node, existingMap = {}, contextStack = []) => {
    const { contextStack: transformedContext, node: transformedNode } = transformNode(contextStack, node)
    const { topLevel: { key, tag, ...topLevelRest }, appearance: { contents } } = pullProperties(transformedNode)
    if (!key || !tag) {
        return existingMap
    }
    const firstPassMap = addElement(
        existingMap,
        {
            contextStack: transformedContext,
            node: {
                ...transformedNode,
                contents: []
            }
        })
    const updatedContextStack = [
        ...transformedContext,
        {
            key,
            tag,
            ...topLevelRest,
            index: (firstPassMap[key]?.appearances || []).length - 1
        }
    ]
    const secondPassMap = (contents || []).reduce((previous, node) => (normalize(node, previous, updatedContextStack)), firstPassMap)
    return secondPassMap
}

export default normalize
