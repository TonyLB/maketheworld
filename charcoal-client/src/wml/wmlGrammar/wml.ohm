WorldMarkupLangage {

    //
    // This section defines the tags that make up the WML language
    //

    WMLFileContents = AssetExpression
            | StoryExpression
            | CharacterExpression

    CharacterExpression = TagExpression<"Character", "key", ("player" | "fileName" | "zone" | "subFolder"), none, none, CharacterLegalContents, none>
    CharacterLegalContents = ImageExpression
            | LiteralNameExpression
            | PronounsExpression
            | LiteralValueTag<"OneCoolThing">
            | LiteralValueTag<"Outfit">
            | LiteralValueTag<"FirstImpression">

    PronounsExpression = TagExpression<"Pronouns", none, ("subject" | "object" | "possessive" | "adjective" | "reflexive"), none, none, none, none>
            | TagSelfClosing<"Pronouns", none, ("subject" | "object" | "possessive" | "adjective" | "reflexive"), none, none, none>

    StoryExpression = TagExpression<"Story", "key", ("fileName" | "zone" | "subFolder" | "player"), none, "instance", AssetLegalContents, none>

    AssetExpression = TagExpression<"Asset", "key", ("fileName" | "zone" | "subFolder" | "player"), none, none, AssetLegalContents, none>
    AssetLegalContents = ImportExpression
            | VariableExpression
            | ComputedExpression
            | ActionExpression
            | LayerExpression
            | ConditionExpression
            | MapExpression
            | RoomExpression
            | ExitExpression
            | FeatureExpression
            | LiteralNameExpression
            | string

    MapExpression = TagExpression<"Map", "key", none, none, none, MapLegalContents, none>
    MapLegalContents = MapRoomExpression
            | NameExpression
            | ImageExpression

    ImageExpression = TagExpression<"Image", "key", "fileURL", none, none, none, none>
            | TagSelfClosing<"Image", "key", "fileURL", none, none, none>

    ImportExpression = TagExpression<"Import", "from", none, none, none, ImportLegalContents, none>
    ImportLegalContents = UseExpression

    UseExpression = TagExpression<"Use", ("key" | "as"), "type", none, none, none, none>
            | TagSelfClosing<"Use", ("key" | "as"), "type", none, none, none>

    DependencyExpression = TagExpression<"Depend", "on", none, none, none, none, none>
            | TagSelfClosing<"Depend", "on", none, none, none, none>

    VariableExpression = TagExpression<"Variable", "key", none, "default", none, none, none>
            | TagSelfClosing<"Variable", "key", none, "default", none, none>

    ComputedExpression = TagExpression<"Computed", "key", none, "src", none, DependencyExpression, none>
            | TagSelfClosing<"Computed", "key", none, "src", none, none>

    ActionExpression = TagExpression<"Action", "key", none, "src", none, none, none>
            | TagSelfClosing<"Action", "key", none, "src", none, none>

    ConditionExpression = TagExpression<"Condition", none, none, "if", none, ConditionLegalContents, none>
    ConditionLegalContents = RoomExpression
            | MapExpression
            | ConditionExpression
            | TextContents
            | LayerExpression
            | ExitExpression
            | FeatureExpression
            | DependencyExpression

    LayerExpression = TagExpression<"Layer", "key", none, none, none, LayerLegalContents, none>
    LayerLegalContents = RoomExpression
            | ConditionExpression
            | ExitExpression

    ExitExpression = TagExpression<"Exit", ("key" | "to" | "from"), none, none, none, string, none>
            | TagSelfClosing<"Exit", ("key" | "to" | "from"), none, none, none, none>

    LinkExpression = TagExpression<"Link", "to", none, none, none, string, spaceCompressor>

    MapRoomExpression = TagExpression<"Room", "key", ("display" | "x" | "y"), none, "global", RoomContents, none>
            | TagSelfClosing<"Room", "key", ("display" | "x" | "y"), none, "global", none>
    RoomExpression = TagExpression<"Room", "key", none, none, "global", RoomContents, none>
            | TagSelfClosing<"Room", "key", "display", none, "global", none>
    RoomContents = DescriptionExpression
            | NameExpression
            | ExitExpression
            | FeatureExpression

    FeatureExpression = TagExpression<"Feature", "key", none, none, "global", FeatureContents, none>
            | TagSelfClosing<"Feature", "key", "display", none, "global", none>
    FeatureContents = DescriptionExpression
            | NameExpression

    DescriptionExpression = TagExpression<"Description", none, "display", none, ("spaceBefore" | "spaceAfter"), DescriptionContents, none>
    DescriptionContents = DescriptionTextContents
            | LinkExpression
            | LineBreakExpression

    LineBreakExpression = "<br" "/>"

    LiteralNameExpression = LiteralValueTag<"Name">
    NameExpression = TagExpression<"Name", none, none, none, ("spaceBefore" | "spaceAfter"), TextContents, none>

    LiteralValueTag<label> = TagExpression<label, none, none, none, none, string, none>

    ValueTag<label> = TagExpression<label, none, none, none, none, TextContents, none>
    DescriptionTextContents = string
    TextContents = string | embeddedJSExpression

    //
    // The following define the ways in which tags can be structured
    //

    TagExpression<tag, tagKeyArgs, tagLiteralArgs, tagProgramArgs, tagBooleanArgs, LegalContents, endSpace> = TagOpen<tag, tagKeyArgs, tagLiteralArgs, tagProgramArgs, tagBooleanArgs> LegalContents* tagClose<tag> #(endSpace)?
    string = (stringText | spaceCompressor)+
    stringText = (~("<" | "{" | space) any | "\\<" | "\\{")+
    legalKey = letter (letter | digit | "_")+
    spaceCompressor = (space)+
    TagSelfClosing<tag, tagKeyArgs, tagLiteralArgs, tagProgramArgs, tagBooleanArgs, endSpace> = "<" tag TagArgument<tagKeyArgs, tagLiteralArgs, tagProgramArgs, tagBooleanArgs>* "/>" #(endSpace)?
    TagOpen<tag, tagKeyArgs, tagLiteralArgs, tagProgramArgs, tagBooleanArgs> = "<" tag TagArgument<tagKeyArgs, tagLiteralArgs, tagProgramArgs, tagBooleanArgs>* ">"
    TagArgument<tagKeyArgs, tagLiteralArgs, tagProgramArgs, tagBooleanArgs> = TagArgumentKey<tagKeyArgs> | TagLiteralArgument<tagLiteralArgs> | TagProgramArgument<tagProgramArgs> | tagBooleanArgument<tagBooleanArgs>
    TagLiteralArgument<tagLiteralArgs> = tagArgumentQuoted<tagLiteralArgs>
    TagProgramArgument<tagProgramArgs> = tagArgumentQuoted<tagProgramArgs> | tagArgumentBracketed<tagProgramArgs>
    tagBooleanArgument<tagBooleanArgs> = tagBooleanArgs &(space | ">" | "/>")
    tagArgumentQuoted<tagLegalArgs> = tagLegalArgs "=\"" tagArgValueQuoted
    TagArgumentKey<tagLegalArgs> = #(tagLegalArgs "=(") legalKey ")"
    tagArgumentBracketed<tagLegalArgs> = tagLegalArgs "={" jsExpression* "}"
    tagArgValueQuoted = ("\\\"" | ~"\"" any)* "\""
    tagClose<tag> = "</" tag ">"
    none = ~">" ">"

    //
    // embeddedJSExpression parses from the starting open bracket through to the
    //   *MATCHING* close bracket, using a subset of JS syntax that concentrates
    //   on all the ways that JS could let you mess up that matching (quotes,
    //   nesting, and template literals figuring high on the list)
    //
    // While we could, technically, validate the entire ES5 JS Syntax (after all,
    // Ohm publishes their schema), it wouldn't gain us any value here, and it
    // has a quite noticeable resource cost:  Compiling that schema and executing
    // it slows the perception Lambda function down to a crawl.  If a need for
    // such pre-processing emerges, look into how to optimize compiling and executing
    // very large Ohm schemata.
    //
    embeddedJSExpression = "{" jsExpression* "}"
    jsExpression = embeddedJSExpression
                    | jsDoubleQuotedString
                    | jsSingleQuotedString
                    | jsTemplateString
                    | jsText
    jsDoubleQuotedString = "\"" jsNoDoubleQuote* "\""
    jsNoDoubleQuote = "\\\"" | (~"\"" any)
    jsSingleQuotedString = "\'" jsNoSingleQuote* "\'"
    jsNoSingleQuote = "\\\'" | (~"\'" any)
    jsTemplateString = "`" (jsNoBackQuote | jsTemplateEvaluation )* "`"
    jsTemplateEvaluation = "${" jsExpression "}"
    jsNoBackQuote = "\\`" | (~("`" | "${") any)
    jsText = ("\\{" | "\\\"" | "\\'" | "\\`" | "\\}" | (~("{" | "\"" | "\'" | "`" | "}") any))+

    // Override Ohm's built-in definition of space.  Cribbed from es5.ohm
    space := whitespace | lineTerminator | comment

	sourceCharacter = any
    whitespace = "\t"
                | "\x0B"    -- verticalTab
                | "\x0C"    -- formFeed
                | " "
                | "\u00A0"  -- noBreakSpace
                | "\uFEFF"  -- byteOrderMark
                | unicodeSpaceSeparator

	unicodeSpaceSeparator = "\u2000".."\u200B" | "\u3000"

    lineTerminator = "\n" | "\r" | "\u2028" | "\u2029"
    lineTerminatorSequence = "\n" | "\r" ~"\n" | "\u2028" | "\u2029" | "\r\n"

    comment = multiLineComment | singleLineComment

    multiLineComment = "/*" (~"*/" sourceCharacter)* "*/"
    singleLineComment = "//" (~lineTerminator sourceCharacter)*

}
