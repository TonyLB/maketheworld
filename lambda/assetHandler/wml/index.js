const fs = require('fs')
const path = require('path')
const ohm = require('ohm-js')
const { compileCode } = require('./compileCode')
const { schema } = require('./semantics/schema')
const { wmlProcessDown, aggregateConditionals, assignExitContext } = require('./semantics/schema/processDown')
const { wmlProcessUp, aggregateErrors, validate } = require('./semantics/schema/processUp')

const wmlSchema = fs.readFileSync(path.join(__dirname, 'wml.ohm'))

const wmlGrammar = ohm.grammar(wmlSchema)

const wmlSemantics = wmlGrammar.createSemantics()
    .addOperation('eval', {
        string(node) {
            return this.sourceString
        },
        EmbeddedJSExpression(open, contents, close) {
            try {
                const evaluation = compileCode(`return (${contents.sourceString})`)({
                    name: 'world'
                })
                return `${evaluation}`
    
            }
            catch(e) {
                return '{#ERROR}'
            }
        },
        _iter(...nodes) {
            return nodes.map((node) => (node.eval())).join('')
        },
        TagExpression(open, contents, close) {
            return `${open.sourceString}${contents.eval()}${close.sourceString}`
        }
    })
    .addOperation('schema', schema)

const tagCondition = (tagList) => ({ tag }) => (tagList.includes(tag))

const flattenToElements = (includeFunction) => (node) => {
    const flattenedNode = includeFunction(node) ? [node] : []
    return node.contents.reduce(
        (previous, node) => ([...previous, ...flattenToElements(includeFunction)(node)]),
        flattenedNode
    )
}

const dbEntries = (match) => {
    const firstPass = wmlSemantics(match).schema()
    const secondPass = wmlProcessDown([
            aggregateConditionals(tagCondition(['Room', 'Exit'])),
            assignExitContext
        ])(firstPass)
    const thirdPass = wmlProcessUp([
            //
            // TODO: Refactor exit validation to assign roomId context as in processDown, then do the calculation (and better error message) knowing all three of
            // to, from and roomId.
            //
            validate(({ tag, to, from }) => ((tag === 'Exit' && !(to && from)) ? ['Exits must have both to and from properties (or be able to derive them from context)'] : [])),
            aggregateErrors
        ])(secondPass)
    const dbSchema = flattenToElements(tagCondition(['Room', 'Exit']))(thirdPass)
    return dbSchema
}

exports.wmlGrammar = wmlGrammar
exports.wmlSemantics = wmlSemantics
exports.dbEntries = dbEntries
