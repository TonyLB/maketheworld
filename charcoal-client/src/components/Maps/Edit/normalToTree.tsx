import { MapTree } from './maps'
import { NormalForm, NormalMap, NormalExit } from '../../../wml/normalize'
import { AssetComponent } from '../../Library/Edit/LibraryAsset'
import { unique } from '../../../lib/lists'

interface NormalToTreeProps {
    MapId: string;
    normalForm: NormalForm;
    rooms: Record<string, AssetComponent>
}

export const normalToTree = ({ MapId, normalForm, rooms }: NormalToTreeProps): MapTree => {
    const map = (normalForm[MapId] || {}) as NormalMap
    const roomItems = (map.appearances || [])
        .filter(({ contextStack }) => (!contextStack.find(({ tag }) => (tag === 'Condition'))))
        .reduce((previous, { rooms }) => ({
            ...previous, 
            ...rooms
        }), {}) as Record<string, { x: number; y: number; location: number[] }>
    const exitPairs = Object.values(normalForm || {})
        .filter(({ tag }) => (tag === 'Exit'))
        .map((item) => (item as NormalExit))
        .reduce<Record<string, string[]>>((previous, { to, from, appearances = [] }) => {
            if (appearances.find(({ contextStack }) => (!contextStack.find(({ tag }) => (tag === 'Condition'))))) {
                return {
                    ...previous,
                    [from]: unique(previous[from] || [], [to])
                }
            }
            return previous
        }, {})
    const exitItems = Object.entries(exitPairs)
        .reduce<MapTree>((previous, [from, toList]) => ([
            ...previous,
            ...(toList.map((to) => ({
                key: `${from}#${to}`,
                item: {
                    name: 'exit',
                    type: 'EXIT' as "EXIT",
                    fromRoomId: from,
                    toRoomId: to,
                    visible: true
                },
                children: []
            })))
        ]), [] as MapTree)
    const tree: MapTree = Object.entries(roomItems || {})
        .reduce<MapTree>((previous, [key, { x = 0, y = 0, location }], index) => {
            return [
                ...previous,
                {
                    key,
                    item: {
                        type: 'ROOM',
                        name: rooms[key]?.name || 'Untitled',
                        location,
                        x,
                        y,
                        roomId: key,
                        visible: true
                    },
                    children: []
                }
            ]
        }, exitItems)
    return [{
        key: 'Default',
        item: {
            type: 'GROUP',
            name: 'Default',
            visible: true
        },
        children: tree
    }]
}

export default normalToTree
