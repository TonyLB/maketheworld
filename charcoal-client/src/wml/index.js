import { compileCode } from './compileCode.js'
import { schema } from './semantics/schema/index.js'
import { wmlProcessDown, aggregateConditionals, assignExitContext } from './semantics/schema/processDown/index.js'
import { wmlProcessUp, aggregateErrors, validate } from './semantics/schema/processUp/index.js'
import wmlGrammar from './wmlGrammar/wml.ohm-bundle.js'

export { wmlGrammar }

export const wmlSemantics = wmlGrammar.createSemantics()
    .addOperation('eval', {
        string(node) {
            return this.sourceString
        },
        EmbeddedJSExpression(open, contents, close) {
            try {
                const evaluation = compileCode(`return (${contents.sourceString})`)({
                    name: 'world'
                })
                return `${evaluation}`
    
            }
            catch(e) {
                return '{#ERROR}'
            }
        },
        _iter(...nodes) {
            return nodes.map((node) => (node.eval())).join('')
        },
        TagExpression(open, contents, close) {
            return `${open.sourceString}${contents.eval()}${close.sourceString}`
        }
    })
    .addOperation('schema', schema)

const tagCondition = (tagList) => ({ tag }) => (tagList.includes(tag))

const flattenToElements = (includeFunction) => (node) => {
    const flattenedNode = includeFunction(node) ? [node] : []
    return node.contents.reduce(
        (previous, node) => ([...previous, ...flattenToElements(includeFunction)(node)]),
        flattenedNode
    )
}

const mergeToRooms = (elements) => {
    const shallowEqual = (listA, listB) => {
        if (listA.length !== listB.length) {
            return false
        }
        return !listA.find((condition, index) => (listB[index] !== condition))
    }
    //
    // TODO: Consider replacing render and name lists with DRY elements that manage the
    // complications of conditional lists internally
    //
    const reduceInRoomContext = ({ render: previousRender = [], exits: previousExits = [], name: previousName = [] }, element) => {
        const mergeRender = (previous = { render: [] }, { display, render }) => {
            //
            // TODO: Handle modes of display other than replace
            //
            return { display, render }
        }
        const mergeName = (previous = { render: [] }, { display, name }) => {
            //
            // TODO: Handle modes of display other than replace
            //
            return { display, name }
        }
        const mergeExit = (previous = {}, exit) => ({ exits: [
            ...(previous?.exits || []).filter((exitProbe) => (exitProbe.to !== exit.to)),
            {
                to: exit.to,
                name: exit.name ? exit.name.trim() : undefined
            }
        ]})
        const mergeSameConditions = (mergeFunction) => (previous, { conditions, ...rest }) => {
            const mergeCandidate = previous.slice(-1)[0]
            if (mergeCandidate && shallowEqual(mergeCandidate.conditions, conditions)) {
                return [
                    ...previous.slice(0, previous.length - 1),
                    { conditions, ...(mergeFunction(mergeCandidate, rest)) }
                ]
            }
            else {
                return [...previous, { conditions, ...(mergeFunction(undefined, rest)) }]
            }
        }
        const { render = "", name = "" } = element
        switch(element.tag) {
            case 'Room':
                return {
                    render: render ? mergeSameConditions(mergeRender)(previousRender, element) : previousRender,
                    name: name ? mergeSameConditions(mergeName)(previousName, element) : previousName,
                    exits: previousExits
                }
            case 'Exit':
                return {
                    render: previousRender,
                    name: previousName,
                    exits: mergeSameConditions(mergeExit)(previousExits, element)
                }
            default:
                return { render: previousRender, exits: previousExits, name: previousName }
        }
    }
    const roomsById = elements.reduce(
        (previous, element) => {
            const roomId = (element.tag === 'Room' && element.key) || (element.tag === 'Exit' && element.from)
            return {
                ...previous,
                [roomId]: reduceInRoomContext(previous[roomId] || { render: [], exits: [], name: [] }, element)
            }
        }, {}
    )
    return roomsById
}

export const validatedSchema = (match) => {
    const firstPass = wmlSemantics(match).schema()
    const secondPass = wmlProcessDown([
        aggregateConditionals(tagCondition(['Room', 'Exit'])),
        assignExitContext
    ])(firstPass)
    const thirdPass = wmlProcessUp([
        //
        // TODO: Refactor exit validation to assign roomId context as in processDown, then do the calculation (and better error message) knowing all three of
        // to, from and roomId.
        //
        validate(({ tag, to, from }) => ((tag === 'Exit' && !(to && from)) ? ['Exits must have both to and from properties (or be able to derive them from context)'] : [])),
        aggregateErrors
    ])(secondPass)
    return thirdPass
}

export const dbEntries = (schema) => {
    const dbSchema = flattenToElements(tagCondition(['Room', 'Exit']))(schema)
    return mergeToRooms(dbSchema)
}

export const assetRegistryEntries = (schema) => {
    //
    // TODO:  Create a breakdown that returns an element for the Asset details, and then
    // elements (once) for each Room that associates it with that Asset and gives it its
    // global UUID (while providing a mapping back to its scoped ID inside the Asset
    // blueprint)
    //
    const elements = flattenToElements(tagCondition(['Asset', 'Character', 'Room', 'Variable']))(schema)
    return elements.map(({ tag, ...rest }) => {
        const { name, fileName, key, global: isGlobal, importMap, player } = rest
        switch(tag) {
            case 'Asset':
                return {
                    tag,
                    name,
                    fileName,
                    key,
                    player,
                    importMap,
                    zone: rest.zone
                }
            case 'Room':
                return {
                    tag,
                    name,
                    isGlobal,
                    key
                }
            case 'Variable':
                return {
                    tag,
                    key
                }
            case 'Character':
                return {
                    tag,
                    key,
                    fileName,
                    player,
                    Name: rest.Name,
                    Pronouns: rest.Pronouns,
                    FirstImpression: rest.FirstImpression,
                    OneCoolThing: rest.OneCoolThing,
                    Outfit: rest.Outfit
                }
            default:
                return { tag, ...rest }
        }
    })
}
