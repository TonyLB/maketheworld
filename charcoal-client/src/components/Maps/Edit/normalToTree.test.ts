import normalToTree from './normalToTree'
import { NormalForm } from '../../../wml/normalize'

describe('normalToTree', () => {
    it('should convert empty cache to empty tree', () => {
        expect(normalToTree({ MapId: '123', normalForm: {}, rooms: {}})).toEqual([])
    })

    it('should convert rooms and exits to items', () => {
        const testNormalForm: NormalForm = {
            Test: {
                key: 'Test',
                tag: 'Asset',
                appearances: [{
                    contextStack: [],
                    contents: [{
                        key: '123',
                        tag: 'Room',
                        index: 0
                    },
                    {
                        key: '456',
                        tag: 'Room',
                        index: 0
                    },
                    {
                        key: '789',
                        tag: 'Room',
                        index: 0
                    },
                    {
                        key: 'TestMap',
                        tag: 'Map',
                        index: 0
                    }]
                }]
            },
            TestMap: {
                tag: 'Map',
                key: 'TestMap',
                appearances: [{
                    contextStack: [{ key: 'Test', tag: 'Asset', index: 0 }],
                    rooms: {
                        '123': {
                            x: 200,
                            y: 150
                        }
                    },
                    contents: [{
                        tag: 'Image',
                        key: 'ImageTest',
                        index: 0
                    },
                    {
                        tag: 'Room',
                        key: '123',
                        index: 1
                    }]
                }]
            },
            '123': {
                key: '123',
                tag: 'Room',
                appearances: [{
                    contextStack: [{ key: 'Test', tag: 'Asset', index: 0 }],
                    name: 'Vortex',
                    contents: []
                },
                {
                    contextStack: [
                        { key: 'Test', tag: 'Asset', index: 0 },
                        { key: 'TestMap', tag: 'Map', index: 0 }
                    ],
                    contents: []
                }]
            },
            '456': {
                key: '456',
                tag: 'Room',
                appearances: [{
                    contextStack: [{ key: 'Test', tag: 'Asset', index: 0 }],
                    name: 'Welcome Room',
                    contents: [
                        { key: '456#123', tag: 'Exit', index: 0 },
                        { key: '456#789', tag: 'Exit', index: 0 }
                    ]
                },
                {
                    contextStack: [
                        { key: 'Test', tag: 'Asset', index: 0 },
                        { key: 'TestMap', tag: 'Map', index: 0 }
                    ],
                    contents: []
                }]
            },
            '456#123': {
                key: '456#123',
                tag: 'Exit',
                to: '123',
                from: '456',
                appearances: [{
                    contextStack: [{ key: 'Test', tag: 'Asset', index: 0 }, { key: '456', tag: 'Room', index: 0 }],
                    contents: []
                }]
            },
            '456#789': {
                key: '456#789',
                tag: 'Exit',
                to: '789',
                from: '456',
                appearances: [{
                    contextStack: [{ key: 'Test', tag: 'Asset', index: 0 }, { key: '456', tag: 'Room', index: 0 }],
                    contents: []
                }]
            },
            '789': {
                key: '123',
                tag: 'Room',
                appearances: [{
                    contextStack: [{ key: 'Test', tag: 'Asset', index: 0 }],
                    name: 'None',
                    contents: []
                },
                {
                    contextStack: [
                        { key: 'Test', tag: 'Asset', index: 0 },
                        { key: 'TestMap', tag: 'Map', index: 0 }
                    ],
                    contents: []
                }]
            },          
        }

        const roomAssetFromName = (name: string) => ({
            localName: '',
            defaultName: name,
            name: name,
            localRender: [],
            defaultRender: [],
            render: []
        })
        expect(normalToTree({
            MapId: 'TestMap',
            normalForm: testNormalForm,
            rooms: {
                '123': roomAssetFromName('Vortex'),
                '456': roomAssetFromName('Welcome Room'),
                '789': roomAssetFromName('None')
            }
        })).toMatchSnapshot()
    })
})