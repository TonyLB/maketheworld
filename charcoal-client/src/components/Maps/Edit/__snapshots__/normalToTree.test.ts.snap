// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`normalToTree should convert empty cache to empty tree 1`] = `
Array [
  Object {
    "children": Array [],
    "item": Object {
      "name": "Default",
      "type": "GROUP",
      "visible": true,
    },
    "key": "Default",
  },
]
`;

exports[`normalToTree should convert rooms and exits to items 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [],
        "item": Object {
          "fromRoomId": "456",
          "name": "exit",
          "toRoomId": "123",
          "type": "EXIT",
          "visible": true,
        },
        "key": "456#123",
      },
      Object {
        "children": Array [],
        "item": Object {
          "fromRoomId": "456",
          "name": "exit",
          "toRoomId": "789",
          "type": "EXIT",
          "visible": true,
        },
        "key": "456#789",
      },
      Object {
        "children": Array [],
        "item": Object {
          "location": Array [
            0,
            0,
            0,
          ],
          "name": "Vortex",
          "roomId": "123",
          "type": "ROOM",
          "visible": true,
          "x": 200,
          "y": 150,
        },
        "key": "123",
      },
    ],
    "item": Object {
      "name": "Default",
      "type": "GROUP",
      "visible": true,
    },
    "key": "Default",
  },
]
`;
